3648 Change "DVN" to default value "producer" and only write the firs…

JingMa87 opened the pull request:
…t geoBndBox element with four geo coordinates.

**What this PR does / why we need it**: Improves the DDI XML according to the XSD

**Which issue(s) this PR closes**: 3648 (doesn't close it, but it's a partial fix)

Closes nothing yet, it's an improvement

**Special notes for your reviewer**: It's an improvement in the DDI XML response, but more is needed in the future.

**Suggestions on how to test this**: Add multiple Geographic Bounding Boxes to your dataset, one with only 3 coordinates, and two with four coordinates. Only the first box with four coordinates should be in the DDI XML.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**: No

**Is there a release notes update needed for this change?**: Yes

**Additional documentation**:

jggautier commented:
Once this PR is merged, will https://github.com/IQSS/dataverse/issues/3648 be reopened? I saw that you wrote that this PR "doesn't close it, but it's a partial fix," and I agree.

It makes sense to me that Dataverse should only allow one set of bounding boxes per dataset. The field is defined as "the largest geographic extent of the Dataset's geographic coverage." And it's true that DDI Codebook allows only one set of geographic bounding boxes in each <sumDscr> element. But it also allows multiple <sumDscr> elements in one DDI document (although I'm not sure why DDI Codebook allows multiple <sumDscr> elements). As part of https://github.com/IQSS/dataverse/issues/3648 I wrote that if it really did make sense to let depositors add multiple bounding box sets, technically multiple <sumDscr> elements could be used for each set:

```
<stdyInfo>
	...
	<sumDscr>
		<timePdr></timePdr>
		...
		<geoBndBox>	
			<westBL>10</westBL><!-- Set one -->
			<eastBL>20</eastBL>
			<southBL>30</southBL>
			<northBL>40</northBL>
			...
			<dataKind></dataKind>
		</geoBndBox>
	</sumDscr>
	<sumDscr>
		<geoBndBox>
			<westBL>50</westBL><!-- Set two -->
			<eastBL>60</eastBL>
			<southBL>80</southBL>
			<northBL>70</northBL>
		</geoBndBox>
	</sumDscr>
</stdyInfo>
```

Like @pdurbin [wrote earlier](https://github.com/IQSS/dataverse/issues/7091#issuecomment-659615995), the Geospatial Discovery Working Group could be a good place to get input about these bounding box questions.

There's also a [DDI Codebook working group](https://ddi-alliance.atlassian.net/wiki/spaces/DDI4/pages/929792030/DDI%2BCodebook%2BDevelopment%2BWork) helping design the next version of DDI Codebook, and if it's discovered that depositors should be able to add more than one bounding box, that should be a good place to ask if using multiple <sumDscr> elements for each bounding box would be a misuse of the element. I'm in that working group and a few others from the Dataverse community are too but I haven't had time to bring these issues up unfortunately.

mheppler commented:
@jggautier if you believe that issue should remain open, you can edit the PR description to remove the "Closes" keyword.

I also added a comment references this PR to other older DDI XML issues I found in the backlog.

JingMa87 commented:
> @jggautier if you believe that issue should remain open, you can edit the PR description to remove the "Closes" keyword.
> 
> I also added a comment references this PR to other older DDI XML issues I found in the backlog.

I removed the Closes keyword and will take a look at the issues you referenced.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/export/ddi/DdiExportUtil.java:
```diff
@@ -328,7 +328,7 @@ private static void writeDocDescElement (XMLStreamWriter xmlw, DatasetDTO datase
     private static void writeVersionStatement(XMLStreamWriter xmlw, DatasetVersionDTO datasetVersionDTO) throws XMLStreamException{
         xmlw.writeStartElement("verStmt");
-        writeAttribute(xmlw,"source","DVN"); 
+        writeAttribute(xmlw,"source","producer"); 
```
The string "DVN_3_0" appears in `if (!SOURCE_DVN_3_0.equals(xmlr.getAttributeValue(null, "source"))) {` in ImportDDIServiceBean. It's not an exact match so I don't *think* this will break dataset import from DDI but I thought I'd at least mention it.

JingMa87 replied:
To me it seems like the "DVN_3_0" is a legacy value and will only still be used if an old Dataverse instance hasn't upgraded yet.

djbrooke commented:
@JingMa87 @jggautier before we move this to QA, do you have any thoughts about @pdurbin's comments?

pdurbin commented:
One last think to mention is that I brought up in tech hours today the idea of changing "source" from "DVN" to something more reasonable. No one objected and it sounds like it's a well known problem that "DVN" doesn't validate.

jggautier commented:
I agree about creating a new issue for the bounding box changes. Hopefully the `<verStmt>` change is easy to make. The bounding box change needs some more discussion I think. @JingMa87, I can create a new issue for the bounding box change if you'd like.

So this PR can be the PR for only the `<verStmt>` change, and changes related to the bounding boxes can be removed?

I would use "archive" instead of "producer," mainly because that's what's used for the "distrbtr source". But also in other places in the DDI, when source = "producer", that means the information is coming from the depositor(s) and when source = "archive", that means the information is coming from the repository. The metadata document version is automatically generated by the repository, so maybe that's another argument for using "archive".

So right now the version statement for the metadata document looks like this:

```
<verStmt source="DVN">
    <version date="2020-07-21" type="RELEASED">1</version>
</verStmt>
```
"DVN" isn't a valid value for "source". So this PR changes it to producer:
```
<verStmt source="producer">
    <version date="2020-07-21" type="RELEASED">1</version>
</verStmt>
```

Again, the distribution statement already uses "archive":

```
<distStmt>
    <distrbtr source="archive">QDR Main Collection</distrbtr>
    <distDate>2020-07-06</distDate>
</distStmt>
```
So I think the version statement should also use "archive":

```
<verStmt source="archive">
    <version date="2020-07-21" type="RELEASED">1</version>
</verStmt>
```
What's lost in this change is the "DVN". In the current DDI export, I'm not sure what's being expressed by `<verStmt source="DVN">`. Is it that the source of the version statement is the Dataverse software itself, as opposed to the repository? Is this a holdover from when there was just one "Dataverse Network"? I don't think this information is being used for anything and it's fine that it's lost in the change from this PR.

JingMa87 commented:
- I also agree about separating the issues. @jggautier I already made an issue for the geoBndBox if you don't mind: https://github.com/IQSS/dataverse/issues/7108
- I tried validating the XML with the codebook, but it's not working and is giving me many pre-validation errors. @jggautier Is it still working for you?
- I'll change the value to archive and revert the other changes for the other issue.

JingMa87 commented:
@pdurbin How come other people's commits are in this PR? Did I use a wrong command?

jggautier commented:
Thanks @JingMa87. About validating the XML with the codebook and pre-validation errors, when I test a Codebook XML export that has only the required fields, the only other error I get (besides the "DVN" error) is about the use of `<setAvail/>` `<useStmt/>`. Is that what you meant by pre-validation errors?

pdurbin commented:
@JingMa87 I don't know. Git is hard. 😄  I'm pretty quick to make new branches at any sign of trouble. And it looks like both @jggautier and I would prefer the "source" change to be made in a dedicated branch/pull request.

JingMa87 commented:
@jggautier I actually meant that the imports that the XSD uses were faulty and prevented the validation from happening. But I fixed these in the meantime and all is well. Maybe I'll incorporate an XSD validation in the source code sometime in the future.
@pdurbin I made a separate issue for the DVN problem https://github.com/IQSS/dataverse/issues/7128. While validating the XML I came across another problem, namely the order of some of the elements. This is easy to fix and the issue can be found here https://github.com/IQSS/dataverse/issues/7128. I'll do it in the near future.
